sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala

```diff
@@ -333,6 +333,14 @@ abstract class SchemaPruningSuite
     }
   }
 
+  testSchemaPruning("select explode of nested field of array of struct and " +
+      "all remaining nested fields") {
```

Comment:
Instead of fixing case by case, can we try to find all the possible cases? 

Also, we need to have the unit test cases for these optimizer rules.